graph-builder/src/registry.rs

```diff
@@ -37,24 +46,164 @@ impl Into<cincinnati::Release> for Release {
+            let manifest_kind_future = authenticated_client
+                .has_manifest(&repo, &tag, None)
+                .and_then(move |manifest_kind| match manifest_kind {
+                    Some(manifest_kind) => Ok(manifest_kind),
+                    None => {
+                        Err(format!("{}:{} doesn't have a manifest", &repo, &tag_clone0).into())
+                    }
+                })
+                .inspect(|manifest_kind| {
+                    trace!("manifest_kind: {:?}", manifest_kind);
+                });
+
+            let manifest_future = authenticated_client.get_manifest(&repo, &tag);
+
+            let layer_digests_future =
+                manifest_kind_future
+                    .join(manifest_future)
+                    .map(|(manifest_kind, manifest)| match manifest_kind {
+                        dkregistry::mediatypes::MediaTypes::ManifestV2S1Signed => {
+                            let m: dkregistry::v2::manifest::ManifestSchema1Signed =
+                                serde_json::from_slice(manifest.as_slice())?;
+                            Ok((m.get_layers(), tag))
+                        }
+                        dkregistry::mediatypes::MediaTypes::ManifestV2S2 => {
+                            let m: dkregistry::v2::manifest::ManifestSchema2 =
+                            serde_json::from_slice(manifest.as_slice())?;
+                            Ok((m.get_layers(), tag))
+                        }
+                        _ => Err(format_err!("unknown manifest_kind '{:?}'", manifest_kind)),
+                    });
+
+            layer_digests_future
         })
-    }
-    Ok(metadata)
+        .map_err(|e| format_err!("{}", e))
+        .and_then(|layer_digests_tag| {
+            let (layer_digests, tag) = layer_digests_tag?;
+            trace!("tag: {:?} layer_digests: {:?}", &tag, &layer_digests);
+
+            let layer_digests_mapped_to_releases = layer_digests.iter().map(|layer_digest| {
+                let (registry_host, repo, tag) = (registry_host.clone(), repo.clone(), tag.clone());
+
+                trace!("Downloading layer {}...", &layer_digest);
+                authenticated_client
+                    .get_blob(&repo, &layer_digest)
+                    .map_err(|e| format_err!("{}", e))
+                    .and_then(move |blob| {
+                        trace!("Layer has {} bytes.", blob.len());
+                        let metadata = extract_metadata_from_layer_blob(&blob, "cincinnati.json")?;
+                        let release = Release {
+                            source: format!("{}/{}:{}", &registry_host, &repo, &tag),
+                            metadata: metadata,
+                        };
+                        trace!("Found release '{:?}'", release);
+                        Ok(release)
+                    })
+            });
+
+            let futures: Vec<_> = layer_digests_mapped_to_releases.collect();
```

Comment:
Use turbofish to avoid the extraneous assignment:

```Rust
Ok(layer_digests_mapped_to_releases.collect::<Vec<_>>())
```

Reply:
We should favor readability. That's why I made that suggestion ;)  Otherwise you quickly end up with something that looks like Go, where you read 20 lines of code to figure out that it is doing a trivial operation.